### PR TITLE
TrackerLogic returns modified Contexts

### LINKED ISSUE #304: TrackerLogic should return modified Contexts
The `HandleAnnounce` and `HandleScrape` methods call PreHooks, which update the context given to them and return the updated version. These methods should at the end return the fully updated `context.Context`, so that frontends can pass it to `AfterAnnounce` and `AfterScrape`.

### PR DESCRIPTION
**Pull request by cedricgc**
Closes #304

### CODE REVIEW

File: middleware/middleware.go [hunk 1]
```diff
@@ -49,20 +49,20 @@ type Logic struct {
 }
 
 // HandleAnnounce generates a response for an Announce.
-func (l *Logic) HandleAnnounce(ctx context.Context, req *bittorrent.AnnounceRequest) (resp *bittorrent.AnnounceResponse, err error) {
+func (l *Logic) HandleAnnounce(ctx context.Context, req *bittorrent.AnnounceRequest) (context context.Context, resp *bittorrent.AnnounceResponse, err error) {
```
Comment (cedricgc) [inline middleware/middleware.go]: Compiler doesn't allow the same name for return variable

Comment (jzelinskie) [inline middleware/middleware.go]: maybe we could call them `prevCtx` and `nextCtx`?

File: frontend/frontend.go [hunk 2]
```diff
@@ -11,14 +11,14 @@ import (
 // after the response has been delivered to the client.
 type TrackerLogic interface {
 	// HandleAnnounce generates a response for an Announce.
```
Comment (mrd0ll4r) [inline frontend/frontend.go]: Add a line about what this returns in which case, something like "returns the updated context, the generated AnnounceResponse and no error on success, ...".

File: frontend/frontend.go [hunk 3]
```diff
@@ -11,14 +11,14 @@ import (
...
-	HandleAnnounce(context.Context, *bittorrent.AnnounceRequest) (*bittorrent.AnnounceResponse, error)
+	HandleAnnounce(context.Context, *bittorrent.AnnounceRequest) (context.Context, *bittorrent.AnnounceResponse, error)
 
 	// AfterAnnounce does something with the results of an Announce after it
 	// has been completed.
 	AfterAnnounce(context.Context, *bittorrent.AnnounceRequest, *bittorrent.AnnounceResponse)
 
 	// HandleScrape generates a response for a Scrape.
```
Comment (mrd0ll4r) [inline frontend/frontend.go]: Same here :)

File: frontend/frontend.go (hunk 2 above)
Comment (cedricgc) [inline frontend/frontend.go]: Returns the updated context, the generated AnnounceResponse and no error on success; nil and error on failure.

File: frontend/frontend.go [hunk 4]
```diff
@@ -10,15 +10,19 @@ import (
 // response from a parsed request, and (2) asynchronously observe anything
 // after the response has been delivered to the client.
 type TrackerLogic interface {
-	// HandleAnnounce generates a response for an Announce.
-	HandleAnnounce(context.Context, *bittorrent.AnnounceRequest) (*bittorrent.AnnounceResponse, error)
+	// HandleAnnounce generates a response for an Announce. Returns the updated
```
Comment (mrd0ll4r) [inline frontend/frontend.go]: Ah, sorry about the trouble, but: We usually have at least a newline, typically even an empty line before the "Returns ...", like so:

```go
// HandleAnnounce generates a response for an Announce.
//
// Returns the updated context, the generated AnnounceResponse and no
// error on success; nil and error on failure.
```
(I don't know if that fits within the line-length limit, only have a browser at hand.)

File: middleware/middleware.go (hunk 1 above)
Comment (mrd0ll4r) [inline middleware/middleware.go]: You could even use the blank identifier `_` to name the return argument.

**Comment by cedricgc** [comment]
will do when I am back home

**Comment by cedricgc** [comment]
oh wait need to add the //

**Comment by mrd0ll4r** [comment]
Alright, code looks good now! Only thing left is to squash this into one commit with a nice name and description :)

File: frontend/udp/frontend.go [hunk 5]
```diff
@@ -286,16 +286,17 @@ func (t *Frontend) handleRequest(r Request, w ResponseWriter) (actionName string
 		af = new(bittorrent.AddressFamily)
 		*af = req.IP.AddressFamily
 
+		var ctx context.Context
```
Comment (jzelinskie) [inline frontend/udp/frontend.go]: context.Background()

File: frontend/udp/frontend.go [hunk 6]
```diff
@@ -319,16 +320,17 @@ func (t *Frontend) handleRequest(r Request, w ResponseWriter) (actionName string
 		af = new(bittorrent.AddressFamily)
 		*af = req.AddressFamily
 
+		var ctx context.Context
```
Comment (jzelinskie) [inline frontend/udp/frontend.go]: context.Background()

File: frontend/udp/frontend.go (hunk 5 above)
Comment (mrd0ll4r) [inline frontend/udp/frontend.go]: I don't think so - this is the same as with `resp` on the next line. They're both gonna be set assigned to in line 291.

Comment (cedricgc) [inline frontend/udp/frontend.go]: I did it to match resp; I can change the assignment to set the value on declaration.

Comment (jzelinskie) [inline frontend/udp/frontend.go]: Yeah, this was a nitpick, but i don't think it really matters at all.

**Pull request merged**